src/clr-angular/utils/focus-trap/focus-trap.directive.ts

```diff
@@ -13,50 +13,99 @@ import {
   Injector,
   OnDestroy,
   PLATFORM_ID,
+  Renderer2,
 } from '@angular/core';
 
 import { FocusTrapTracker } from './focus-trap-tracker.service';
 
 @Directive({ selector: '[clrFocusTrap]' })
 export class FocusTrapDirective implements AfterViewInit, OnDestroy {
-  private _previousActiveElement: HTMLElement;
-  /* tslint:disable-next-line:no-unused-variable */
+  private previousActiveElement: any;
   private document: Document;
 
+  private topReboundEl: any;
+  private bottomReboundEl: any;
+
   constructor(
-    public elementRef: ElementRef,
-    injector: Injector,
+    private el: ElementRef,
+    private injector: Injector,
     private focusTrapsTracker: FocusTrapTracker,
+    private renderer: Renderer2,
     @Inject(PLATFORM_ID) private platformId: Object
   ) {
-    this.document = injector.get(DOCUMENT);
+    this.document = this.injector.get(DOCUMENT);
     this.focusTrapsTracker.current = this;
   }
 
   @HostListener('document:focusin', ['$event'])
   onFocusIn(event: any) {
-    const nativeElement: HTMLElement = this.elementRef.nativeElement;
+    const nativeElement: HTMLElement = this.el.nativeElement;
 
-    if (this.focusTrapsTracker.current === this && !nativeElement.contains(event.target)) {
+    if (this.focusTrapsTracker.current === this && event.target && !nativeElement.contains(event.target)) {
       nativeElement.focus();
     }
   }
 
-  ngAfterViewInit() {
-    if (isPlatformBrowser(this.platformId)) {
-      this._previousActiveElement = <HTMLElement>document.activeElement;
-      const nativeElement: HTMLElement = this.elementRef.nativeElement;
-      nativeElement.setAttribute('tabindex', '0');
+  private createFocusableOffScreenEl(nodeName: string, optionalClass?: string): any {
+    const offScreenSpan = this.renderer.createElement(nodeName);
+    this.renderer.setAttribute(offScreenSpan, 'tabindex', '0');
+    this.renderer.addClass(offScreenSpan, 'offscreen-focus-rebounder');
+
+    if (optionalClass) {
+      this.renderer.addClass(offScreenSpan, optionalClass);
+    }
+
+    return offScreenSpan;
+  }
+
+  private addReboundEls() {
+    // We will add these focus rebounding elements only in the following conditions:
+    // 1. It should be running inside browser platform as it accesses document.body element
+    // 2. We should add them more than once. Hence, we are counting a number of focus trappers
+    //    and only add on the first focus trapper.
+
+    if (this.focusTrapsTracker.nbFocusTrappers === 1) {
+      this.topReboundEl = this.createFocusableOffScreenEl('span', 'before-focus-trap');
+      this.bottomReboundEl = this.createFocusableOffScreenEl('span', 'after-focus-trap');
+
+      // Add reboundBeforeTrapEl to the document body as the first child
+      const bodyFirstChild = this.document.body.firstChild;
+      if (bodyFirstChild) {
```

Comment:
This will always be true, since the body has to have at least one element for Angular. Don't need to capture it as a variable I think.

```
this.renderer.insertBefore(this.document.body, this.topReboundEl, this.document.body.firstChild);
```

Reply:
You are right. I will update this. Thanks

Reply:
Done